util/GenerateDocExamples.php

```diff
@@ -0,0 +1,203 @@
+    checkIfCodeHasValidSyntax($code);
+
+    $exampleFile = sprintf("%s/docs/examples/%s.asciidoc", dirname(__DIR__), $source['digest']);
+    file_put_contents($exampleFile, $head . "\n" . $code . "----\n");
+    printf("File generated: %s.asciidoc\n", $source['digest']);
+    $parsed++;
+}
+$end = microtime(true);
+
+// Count only the digests to prevent double count in code examples
+$tot = count($digests);
+
+printf("\nLanguage statistics:\n");
+foreach ($langs as $lang => $num) {
+    printf("%-8s: %4d (%.2f%% completed)\n", $lang, $num, (100/$tot)*$num);
+}
+printf("\n");
+
+printf("Total source examples: %d\n", $tot);
+printf("Generated %d source examples in %.3f seconds\n", $parsed, $end - $start);
+
+// END 
+
+/**
+ * Generate the client source code from the parsed_source field
+ */
+function getClientSourceCode(array $parsedSource): string
+{
+    $tab4 = str_repeat(' ', 4);
+    $code = '';
+    foreach ($parsedSource as $source) {
+        if (!empty($source['params']) || !empty($source['body'])) {
+            $code .= '$params = [' . "\n";
+            $code .= prettyPrintArray($source['params'], 4);
+            if (isset($source['body'])) {
+                $code .= sprintf("%s'body' => [\n", $tab4);
+                $code .= prettyPrintArray($source['body'], 8);
+                $code .= sprintf("%s],\n", $tab4);
+            }
+            $code .= "];\n";
+        }
+        $code .= '$response = $client->' . normalizeApiName($source['api']);
+        $code .= empty($source['params']) && empty($source['body']) ? '();' : '($params);';
+        $code .= "\n";
+    }
+    return $code;
+}
+
+/**
+ * Print an associative array as source code
+ * Note: it removes the number keys in scalar array
+ */
+function prettyPrintArray(array $input, int $space): string
+{
+    $output = '';
+    $tab = str_repeat(' ', $space);
+    foreach ($input as $key => $value) {
+        if (is_int($key)) {
+            $output .= sprintf("%s", $tab);
+        } else {
+            $output .= sprintf("%s'%s' => ", $tab, $key);
+        }
+        if (is_array($value)) {
+            $output .= "[\n";
+            $output .= prettyPrintArray($value, $space + 4);
+            $output .= sprintf("%s],\n", $tab);
+        } else {
+            if (is_string($value)) {
+                $value = "'" . str_replace("'", "\'", $value) . "'";
+            } if (is_bool($value)) {
+                $value = $value ? 'true' : 'false';
+            }
+            $output .= sprintf("%s,\n", $value);
+        }
+    }
+    return $output;
+}
+
+/**
+ * Normalize the api name for invoking the equivalent endpoint
+ */
+function normalizeApiName(string $api): string
+{
+    $result = str_replace('_', '', lcfirst(ucwords($api, '_')));
+    return str_replace('.', '()->', $result);
+}
+
+/**
+ * Check if the generated code has a valid PHP syntax using Elasticsearch\Client
+ */
+function checkIfCodeHasValidSyntax(string $code): void
```

Comment:
How about you generate PHPUnit tests instead?

```php
class DocExamplesTest extends TestCase {

    public function test<Insert-digest-here>()
    {
        $client = $this->getClient();
        //
        // <paste $code here>
        //
        $this->assertTrue(true);  // if no exception is thrown the test is good
    }

// more tests

}
```

Reply:
I need a way to test the generated code on the fly, that's why I used this approach. I can definitely write a unit test helper to iterate on all the `.asciidoc` files and run a test on each digest. This will require a valid Elasticsearch instance running with some testing data on it. We need to check on the testing environment of https://github.com/elastic/elasticsearch.

Reply:
> I need a way to test the generated code on the fly, that's why I used this approach
That's the part was unsure about, ++ on your approach